internal/controller/namespaced/object/syncer.go

```diff
@@ -110,8 +126,69 @@ func (s *SSAResourceSyncer) GetDesiredState(ctx context.Context, obj *v1alpha1.O
 
 // SyncResource syncs the supplied object by using server-side apply to apply.
 func (s *SSAResourceSyncer) SyncResource(ctx context.Context, obj *v1alpha1.Object, desired *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	// first, upgrade managed fields to SSA manager if needed
+	err := s.maybeUpgradeFieldManagers(ctx, obj)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot upgrade field managers")
+	}
 	if err := s.client.Patch(ctx, desired, client.Apply, client.ForceOwnership, client.FieldOwner(ssaFieldOwner(obj.GetName()))); err != nil {
 		return nil, errors.Wrap(CleanErr(err), errCreateObject)
 	}
 	return desired, nil
 }
+
+// needSSAFieldManagerUpgrade checks the given k8s resource has legacy CSA field
+// managers in the managed field entries.
+func (s *SSAResourceSyncer) needSSAFieldManagerUpgrade(accessor metav1.Object) bool {
+	if accessor == nil {
+		return false
+	}
+	mfes := accessor.GetManagedFields()
+	for _, mfe := range mfes {
+		if mfe.Operation == metav1.ManagedFieldsOperationUpdate && s.legacyCSAFieldManagers.Has(mfe.Manager) {
+			return true
+		}
+	}
+	return false
+}
+
+// maybeUpgradeFieldManagers upgrades managed field entries of the managed k8s resource
+// to the SSA field owner.
+func (s *SSAResourceSyncer) maybeUpgradeFieldManagers(ctx context.Context, obj *v1alpha1.Object) error {
+	current, err := parseStatus(obj)
+	if err != nil {
+		return errors.Wrap(err, "cannot get the current state of the object")
+	}
+	if current == nil || !s.needSSAFieldManagerUpgrade(current) {
+		return nil
+	}
+	// this generates a raw json patch that transfers only the specified
+	// CSA field managers to the SSA field manager. Other field managers
+	// that might exist are not modified.
+	// The generated patch includes resourceVersion, i.e. uses optimistic
+	// locking, so if the object was changed between the last observation
+	// and the patch, it will be safely rejected, then requeued.
+	mfUpgradePatch, err := csaupgrade.UpgradeManagedFieldsPatch(current, s.legacyCSAFieldManagers, ssaFieldOwner(obj.GetName()))
```

Comment:
This function [may return](https://github.com/kubernetes/client-go/blob/54601aa99ffde1d9a79128043b852f41fad21415/util/csaupgrade/upgrade.go#L120) a `nil` patch if nothing needs to be done. So, it is a good idea to handle nil patches, e.g. return nil and not call client.patch.